geomstats/learning/frechet_mean.py

```diff
@@ -0,0 +1,348 @@
+    var = gs.to_ndarray(var, to_ndim=2, axis=1)
+    return var
+
+
+def linear_mean(points, weights=None):
+    """Compute the frechet mean.
+
+    The Frechet mean of (weighted) points computed with the
+    Euclidean metric is the weighted average of the points
+    in the Euclidean space.
+
+    Parameters
+    ----------
+    points: array-like, shape=[n_samples, dimension]
+
+    weights: array-like, shape=[n_samples, 1], optional
+
+    Returns
+    -------
+    mean: array-like, shape=[1, dimension]
+    """
+    if isinstance(points, list):
+        points = gs.vstack(points)
+    points = gs.to_ndarray(points, to_ndim=2)
+    n_points = gs.shape(points)[0]
+
+    if isinstance(weights, list):
+        weights = gs.vstack(weights)
+    elif weights is None:
+        weights = gs.ones((n_points, ))
+
+    weighted_points = gs.einsum('n,nj->nj', weights, points)
+    mean = (gs.sum(weighted_points, axis=0) / gs.sum(weights))
+    mean = gs.to_ndarray(mean, to_ndim=2)
+    return mean
+
+
+def _default_gradient_descent(points, metric, weights,
+                              n_max_iterations, point_type, epsilon, verbose):
+
+    def while_loop_cond(iteration, mean, var, sq_dist):
+        result = ~gs.logical_or(
+            gs.isclose(var, 0.),
+            gs.less_equal(sq_dist, epsilon * var))
+        return result[0, 0] or iteration == 0
+
+    def while_loop_body(iteration, mean, var, sq_dist):
+
+        logs = metric.log(point=points, base_point=mean)
+
+        tangent_mean = gs.einsum('nk,nj->j', weights, logs)
+
+        tangent_mean /= sum_weights
+
+        estimate_next = metric.exp(
+            tangent_vec=tangent_mean,
+            base_point=mean)
+
+        sq_dist = metric.squared_dist(estimate_next, mean)
+        sq_dists_between_iterates.append(sq_dist)
+
+        var = variance(
+            points=points,
+            weights=weights,
+            metric=metric,
+            base_point=estimate_next)
+
+        mean = estimate_next
+        iteration += 1
+        return [iteration, mean, var, sq_dist]
+
+    if point_type == 'vector':
+        points = gs.to_ndarray(points, to_ndim=2)
+    if point_type == 'matrix':
+        points = gs.to_ndarray(points, to_ndim=3)
+    n_points = gs.shape(points)[0]
+
+    if weights is None:
+        weights = gs.ones((n_points, 1))
+
+    weights = gs.array(weights)
+    weights = gs.to_ndarray(weights, to_ndim=2, axis=1)
+
+    sum_weights = gs.sum(weights)
+
+    mean = points[0]
+    if point_type == 'vector':
+        mean = gs.to_ndarray(mean, to_ndim=2)
+    if point_type == 'matrix':
+        mean = gs.to_ndarray(mean, to_ndim=3)
+
+    if n_points == 1:
+        return mean
+
+    sq_dists_between_iterates = []
+    iteration = 0
+    sq_dist = gs.array([[0.]])
+    var = gs.array([[0.]])
+
+    last_iteration, mean, var, sq_dist = gs.while_loop(
+        lambda i, m, v, sq: while_loop_cond(i, m, v, sq),
+        lambda i, m, v, sq: while_loop_body(i, m, v, sq),
```

Comment:
How did you get lambda functions to pass flake8??

Reply:
haha, no idea!